modules/src/applications/transfer/context.rs

```diff
@@ -0,0 +1,348 @@
+
+    /// Returns true iff the store contains a `DenomTrace` entry for the specified `HashedDenom`.
+    fn has_denom_trace(&self, hashed_denom: &HashedDenom) -> bool {
+        self.get_denom_trace(hashed_denom).is_some()
+    }
+
+    /// Get the denom trace associated with the specified hash in the store.
+    fn get_denom_trace(&self, denom_hash: &HashedDenom) -> Option<DenomTrace>;
+}
+
+pub trait BankKeeper {
+    type AccountId: ToString;
+
+    /// This function should enable sending ibc fungible tokens from one account to another
+    fn send_coins(
+        &mut self,
+        from: &Self::AccountId,
+        to: &Self::AccountId,
+        amt: &IbcCoin,
+    ) -> Result<(), Ics20Error>;
+
+    /// This function to enable minting ibc tokens in a module
+    fn mint_coins(&mut self, module: &Self::AccountId, amt: &IbcCoin) -> Result<(), Ics20Error>;
+
+    /// This function should enable burning of minted tokens
+    fn burn_coins(&mut self, module: &Self::AccountId, amt: &IbcCoin) -> Result<(), Ics20Error>;
+
+    /// This function should enable transfer of tokens from the ibc module to an account
+    fn send_coins_from_module_to_account(
+        &mut self,
+        module: &Self::AccountId,
+        to: &Self::AccountId,
+        amt: &IbcCoin,
+    ) -> Result<(), Ics20Error>;
+
+    /// This function should enable transfer of tokens from an account to the ibc module
+    fn send_coins_from_account_to_module(
+        &mut self,
+        from: &Self::AccountId,
+        module: &Self::AccountId,
+        amt: &IbcCoin,
+    ) -> Result<(), Ics20Error>;
+}
+
+pub trait BankReader {
+    type AccountId: ToString + TryFrom<Signer>;
+
+    /// Returns true if the specified account is not allowed to receive funds and false otherwise.
+    fn is_blocked_account(&self, account: &Self::AccountId) -> bool;
+
+    /// get_transfer_account returns the ICS20 - transfers AccountId.
+    fn get_transfer_account(&self) -> Self::AccountId;
+}
+
+pub trait AccountReader {
+    type AccountId: ToString;
+    type Address;
+
+    fn get_account(&self, address: &Self::Address) -> Option<Self::AccountId>;
+}
+
+/// Captures all the dependencies which the ICS20 module requires to be able to dispatch and
+/// process IBC messages.
+pub trait Ics20Context:
+    Ics20Keeper<AccountId = <Self as Ics20Context>::AccountId>
+    + Ics20Reader<AccountId = <Self as Ics20Context>::AccountId>
+{
+    type AccountId: ToString + TryFrom<Signer>;
+}
+
+fn validate_transfer_channel_params(
+    ctx: &mut impl Ics20Context,
+    order: Order,
+    port_id: &PortId,
+    channel_id: &ChannelId,
+    version: &Version,
+) -> Result<(), Ics20Error> {
+    if channel_id.sequence() > (u32::MAX as u64) {
+        return Err(Ics20Error::chan_seq_exceeds_limit(channel_id.sequence()));
+    }
+
+    if order != Order::Unordered {
+        return Err(Ics20Error::channel_not_unordered(order));
+    }
+
+    let bound_port = ctx.get_port()?;
+    if port_id != &bound_port {
```

Comment:
ICS-20 doesn't have this check (only ibc-go does). Is there a good reason to force chains to only bind the ICS-20 module to a predefined port?

Reply:
I think this is checking if the port-id specified in channel messages matches the port that the ICS-20 module is bound to. AFAIK the core IBC router should only call this module callback if it is registered against this port - that means the `port_id` in the above code is guaranteed to be the port that this module is bound to, and that's a guarantee that core IBC provides. It is however possible that `get_port()` returns a port that is not the one this module is bound to, possibly due to a app dev error.

Reply:
So this check guards against an application development error? If so, makes sense to keep it.

Reply:
I think I was more referring to the fact that modules can bind multiple ports in general, and so theoretically ICS-20 module could be bound to `"transfer"` and `"other_port"`. However, this check prevents us from doing that. I don't know if a chain will ever need that though? The upside of this tradeoff is indeed that it guards against dev error. I don't have a strong preference for keeping/removing it!